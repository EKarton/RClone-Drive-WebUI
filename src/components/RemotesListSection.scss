.remotes-list-section {
  &__header-text {
    padding-bottom: 1rem;
    font-weight: bold;
  }
}
