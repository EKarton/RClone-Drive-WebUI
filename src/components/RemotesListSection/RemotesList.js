import InfoCard from 'components/RemoteCards/InfoCard';
import RemoteCardsList from 'components/RemoteCardsList';

export default function RemotesList({ remotes, onRemoteCardClicked }) {
  const handleClick = (remote) => () => {
    onRemoteCardClicked(remote);
  };

  return (
    <RemoteCardsList>
      {remotes.sort().map((remote) => (
        <InfoCard key={remote} remote={remote} onClick={handleClick(remote)} />
      ))}
    </RemoteCardsList>
  );
}
