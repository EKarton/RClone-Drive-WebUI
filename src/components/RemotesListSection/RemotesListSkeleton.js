import SkeletonCard from 'components/RemoteCards/SkeletonCard';
import RemoteCardsList from 'components/RemoteCardsList';

export default function RemotesListSkeleton() {
  return (
    <RemoteCardsList>
      {Array.from({ length: 4 }, (_, i) => (
        <SkeletonCard key={i} />
      ))}
    </RemoteCardsList>
  );
}
