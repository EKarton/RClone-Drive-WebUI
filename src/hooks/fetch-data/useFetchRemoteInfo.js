import { useCallback } from 'react';
import useFetchData from './useFetchData';

export default function useFetchRemoteInfo(remote) {
  const func = (c, cancelToken) => c.fetchRemoteInfo(remote, { cancelToken });
  return useFetchData(useCallback(func, [remote]));
}
