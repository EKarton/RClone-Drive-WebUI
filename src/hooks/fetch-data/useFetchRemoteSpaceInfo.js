import { useCallback } from 'react';
import useFetchData from './useFetchData';

export default function useFetchRemoteSpaceInfo(remote) {
  const func = (c, cancelToken) => c.fetchRemoteSpaceInfo(remote, { cancelToken });
  return useFetchData(useCallback(func, [remote]));
}
