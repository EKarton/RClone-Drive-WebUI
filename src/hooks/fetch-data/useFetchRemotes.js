import { useCallback } from 'react';
import useFetchData from './useFetchData';

export default function useFetchRemotes() {
  const func = (c, cancelToken) => c.fetchRemotes({ cancelToken });
  return useFetchData(useCallback(func, []));
}
