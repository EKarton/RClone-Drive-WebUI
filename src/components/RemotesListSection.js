import Typography from '@mui/material/Typography';
import RemoteCardList from 'components/RemoteCardList';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';
import { StatusTypes } from 'utils/constants';
import './RemotesListSection.scss';

/**
 * Represents the list of remotes with header
 */
export default function RemotesListSection({ onRemoteCardClicked }) {
  const { status, data: remotes, error } = useFetchRemotes();

  if (status === StatusTypes.LOADING) {
    return null;
  }

  if (status === StatusTypes.ERROR) {
    throw error;
  }

  const handleRemoteCardClicked = (remote) => {
    onRemoteCardClicked(remote);
  };

  return (
    <>
      <Typography
        variant="h7"
        component="div"
        color="text.primary"
        className="remotes-list-section__header-text"
      >
        Your Remotes
      </Typography>
      <RemoteCardList remotes={remotes} onClick={handleRemoteCardClicked} />
    </>
  );
}
