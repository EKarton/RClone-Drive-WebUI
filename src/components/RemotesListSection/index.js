import Typography from '@mui/material/Typography';
import PropTypes from 'prop-types';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';
import { StatusTypes } from 'utils/constants';
import RemotesList from './RemotesList';
import RemotesListSkeleton from './RemotesListSkeleton';

/**
 * Represents the list of remotes with header
 */
export default function RemotesListSection({ onRemoteCardClicked }) {
  const { status, data: remotes, error } = useFetchRemotes();

  if (status === StatusTypes.ERROR) {
    throw error;
  }

  const renderList = () => {
    if (status === StatusTypes.LOADING) {
      return <RemotesListSkeleton />;
    }

    return <RemotesList remotes={remotes} onRemoteCardClicked={onRemoteCardClicked} />;
  };

  return (
    <>
      <Typography variant="h7" component="div" color="text.primary">
        <strong>Your Remotes</strong>
      </Typography>
      {renderList()}
    </>
  );
}

RemotesListSection.propTypes = {
  onRemoteCardClicked: PropTypes.func,
};
