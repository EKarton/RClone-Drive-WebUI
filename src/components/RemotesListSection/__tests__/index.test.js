import useFetchRemoteInfo from 'hooks/fetch-data/useFetchRemoteInfo';
import useFetchRemoteSpaceInfo from 'hooks/fetch-data/useFetchRemoteSpaceInfo';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';
import useRCloneClient from 'hooks/rclone/useRCloneClient';
import { StatusTypes } from 'utils/constants';
import { mockConfigGetResponse } from 'test-utils/mock-responses';
import { mockOperationsAboutResponse } from 'test-utils/mock-responses';
import { mockRemotes } from 'test-utils/mock-responses';
import { customRender, userEvent, screen } from 'test-utils/react';
import RemotesListSection from '..';

jest.mock('hooks/fetch-data/useFetchRemotes');
jest.mock('hooks/fetch-data/useFetchRemoteInfo');
jest.mock('hooks/fetch-data/useFetchRemoteSpaceInfo');
jest.mock('hooks/rclone/useRCloneClient');

describe('RemotesListSection', () => {
  beforeEach(() => {
    useRCloneClient.mockReturnValue({
      emptyTrashCan: jest.fn().mockResolvedValue(),
    });

    useFetchRemotes.mockReturnValue({
      status: StatusTypes.SUCCESS,
      data: mockRemotes.remotes,
      refetchData: jest.fn(),
    });

    useFetchRemoteSpaceInfo.mockReturnValue({
      status: StatusTypes.SUCCESS,
      data: mockOperationsAboutResponse,
      refetchData: jest.fn(),
    });

    useFetchRemoteInfo.mockReturnValue({
      status: StatusTypes.SUCCESS,
      data: mockConfigGetResponse,
      refetchData: jest.fn(),
    });
  });

  it('should match snapshot when api call was successful', async () => {
    const { baseElement } = renderComponent();

    expect(baseElement).toMatchSnapshot();
  });

  it('should throw an error when api call fails', async () => {
    useFetchRemotes.mockReturnValue({
      status: StatusTypes.ERROR,
      error: new Error('Error!'),
    });

    expect(renderComponent).toThrowError();
  });

  it('should match snapshot when api call is loading', async () => {
    useFetchRemotes.mockReturnValue({
      status: StatusTypes.LOADING,
    });

    const { baseElement } = renderComponent();

    expect(baseElement).toMatchSnapshot();
  });

  it('should call onRemoteCardClicked() correctly when user clicks on a card', async () => {
    const view = renderComponent();

    userEvent.click(screen.getByText('googledrive'));

    expect(view.onRemoteCardClicked).toBeCalledWith('googledrive');
  });

  const renderComponent = () => {
    const onRemoteCardClicked = jest.fn();

    const view = customRender(
      <RemotesListSection onRemoteCardClicked={onRemoteCardClicked} />
    );

    view.onRemoteCardClicked = onRemoteCardClicked;
    return view;
  };
});
