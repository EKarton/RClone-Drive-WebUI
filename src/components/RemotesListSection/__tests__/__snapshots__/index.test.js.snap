// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemotesListSection should match snapshot when api call is loading 1`] = `
<body>
  <div>
    <div
      class="MuiTypography-root MuiTypography-h7 css-jab4h0-MuiTypography-root"
    >
      <strong>
        Your Remotes
      </strong>
    </div>
    <div
      class="remote-card-list"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              />
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`RemotesListSection should match snapshot when api call was successful 1`] = `
<body>
  <div>
    <div
      class="MuiTypography-root MuiTypography-h7 css-jab4h0-MuiTypography-root"
    >
      <strong>
        Your Remotes
      </strong>
    </div>
    <div
      class="remote-card-list"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
            >
              googledrive
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              Type: crypt
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              126 GB / 161 GB used, 0 B in trash
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
            >
              icloud
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              Type: crypt
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              126 GB / 161 GB used, 0 B in trash
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
            >
              onedrive
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              Type: crypt
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              126 GB / 161 GB used, 0 B in trash
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
